knowledge_repo/app/auth_providers/oauth2.py

```diff
@@ -0,0 +1,165 @@
+    output = []
+    for variable in variables:
+        value = local_namespace.get(variable, None)
+        if value is None:
+            value = config.get('OAUTH_{}_{}'.format(name, variable).upper(), None)
+        if value is None:
+            value = presets.get(variable, None)
+        output.append(value)
+    return output
+
+
+class OAuth2Provider(KnowledgeAuthProvider):
+    _registry_keys = list(PRESETS.keys())
+
+    def init(self,
+             authorization_url=None,
+             token_url=None,
+             auto_refresh_url=None,
+             scopes=None,
+             user_info_endpoint=None,
+             client_id=None,
+             client_secret=None,
+             user_info_mapping=None,
+             validate=None):
+
+        (self.base_url,
+         self.authorization_url,
+         self.token_url,
+         self.auto_refresh_url,
+         self.scopes,
+         self.user_info_endpoint,
+         self.client_id,
+         self.client_secret,
+         self.user_info_mapping,
+         validate) = _resolve_oauth_config(
+            self.name,
+            locals(),
+            self.app.config,
+            'base_url',
+            'authorization_url',
+            'token_url',
+            'auto_refresh_url',
+            'scopes',
+            'user_info_endpoint',
+            'client_id',
+            'client_secret',
+            'user_info_mapping',
+            'validate'
+        )
+        if validate is not None:
+            self.validate = validate
+
+        redirect_url = self.app.config['SERVER_NAME'] or 'localhost:7000'
+        if self.app.config['APPLICATION_ROOT']:
+            redirect_url = '/'.join((redirect_url, self.app.config['APPLICATION_ROOT']))
+
+        # Import OAuth deps here so we do not have a hard dependency on them
+        from requests_oauthlib import OAuth2Session
+        self.oauth_client = OAuth2Session(
+            client_id=self.client_id,
+            scope=self.scopes,
+            redirect_uri='http://{}/auth/login/{}/authorize'.format(redirect_url, self.name),
+            auto_refresh_url=self.auto_refresh_url
+        )
+
+    def prompt(self):
+        return redirect(self.oauth_client.authorization_url(self.authorization_url)[0])
+
+    def get_endpoint_url(self, endpoint):
+        return urljoin(self.base_url, endpoint)
+
+    def get_user(self):
+        code = request.args.get('code')
+        state = request.args.get('state')
+
+        self.oauth_client.state = state
+        self.oauth_client.fetch_token(self.token_url, client_secret=self.client_secret, code=code)
+
+        return self.extract_user_from_api()
+
+    def extract_user_from_api(self):
+
+        def extract_from_dict(d, key):
+            if isinstance(key, (list, tuple)):
+                if len(key) == 1:
+                    key = key[0]
+                else:
+                    return extract_from_dict(d[key[0]], key[1:])
+            elif isinstance(key, str):
```

Comment:
Pretty sure you want `if isinstance(key, str):` here for the desired behaviour. As is, passing in any tuple or list as key results in `RuntimeError: Invalid key type: (name of last string in list or tuple).`